dashboard/db/migrate/20210311012741_add_indexes_to_programming_expression.rb

```diff
@@ -0,0 +1,6 @@
+class AddIndexesToProgrammingExpression < ActiveRecord::Migration[5.2]
+  def change
+    add_index :programming_expressions, [:name, :category], type: :fulltext
+    add_index :programming_expressions, [:key, :programming_environment_id], unique: true, name: 'key_programming_environment'
```

Comment:
based on  this line from  #39381, https://github.com/code-dot-org/code-dot-org/blob/023987976fd37247986e373e9b2a9dd72fb209e8/dashboard/lib/programming_expression_autocomplete.rb#L6 I would recommend that you swap the order of the fields in this index to be `[:programming_environment_id, :key]`. This will let the line I quoted use this index. I am assuming you are never going to look up the programming expression just by key, because the key only makes sense within the context of a programming environment.